File: JS/insert.js
/**
 * Insertion sort implementation in Javascript.
 * Generates and sorts a list of 10 random integers 1-100.
 * Tracks and displays # of array accesses and comparisons completed during sorting.
 * 
 * Author: Lucas Burns
 * Version: 2017-9-26
 * 
 */

const process = require('./process.js');

/**
 * Sorts a given array of numbers using Insertion sort
 * @param {Array<number>} items 
 */
function sort(items){
	var i = 1;
	while (i < items.length){
		var j = i;
		while (j > 0 && items[j - 1] > items[j]) {
			process.swap(j - 1, j, items);
			accesses += 2;
			comparisons++;
			j--;
		}
		i++;
	}
}

var a = process.populate();

console.log("Original array:		[" + a + "]");
comparisons = 0;
accesses = 0;

sort(a);

console.log("Sorted array:	 	[" + a + "]");
console.log("%d comparisons		%d array accesses.", comparisons, accesses);

File: ruby/insert.rb
require_relative 'process'
include Process

def sort(items)
	i = 1
	while i < items.length do
		j = i
		while j > 0 && items[j - 1] > items[j] do
			$comparisons += 1
			Process.swap(j - 1, j, items)
			$accesses += 5
			j -= 1
		end
		i += 1
	end
end

$comparisons = 0
$accesses = 0

a = Process.populate()

puts("Original array: 	" + a.to_s)

sort(a)

puts("Sorted array: 		" + a.to_s)

puts($comparisons.to_s + " comparisons		" + $accesses.to_s + " acccesses")